text/0000-with-bounds-in-pi-types.md

```diff
@@ -0,0 +1,685 @@
+      P
+      ────────
+      ¬Q
+
+Basically, this states that if two propositions are mutually exclusive (that
+is, not both of them can be true), and one of them is true, the other must be
+false, due to being disjunctive.
+
+This is strictly reductive.
+
+Now let's go funky:
+
+### Cancelation
+
+    AdditiveCancelationRR:
+      a + c = b + c
+      ─────────────
+      a = b
+    AdditiveCancelationLL:
+      c + a = c + b
+      ─────────────
+      a = b
+    AdditiveCancelationRL:
+      a + c = c + b
+      ─────────────
+      a = b
+    AdditiveCancelationLR:
+      c + a = b + c
+      ─────────────
+      a = b
+    MultiplicativeCancelationRR:
+      ac = bc
+      ─────────────
+      a = b
+    MultiplicativeCancelationLL:
+      ca = cb
+      ─────────────
+      a = b
+    MultiplicativeCancelationRL:
+      ac = cb
+      ─────────────
+      a = b
+    MultiplicativeCancelationLR:
+      ca = bc
+      ─────────────
+      a = b
+
+These are all reductive.
+
+### Inequalities
+
+Inequalities are something, we are interested in simplifying. These carry many
+interesting properties, which shall be covered:
+
+    RewriteGeq:
+      a ≥ b
+      ─────
+      b ≤ a
+
+Here, we rewrite greater than or equal to a form of less than or equal.
+
+    RewriteLessThan:
+      a < b
+      ────────
+      a ≤ b
+      ¬(a = b)
+    RewriteGreaterThan:
+      a > b
+      ────────
+      a ≥ b
+      ¬(a = b)
+
+This allows us to rewrite less than (`<`) and greater than (`>`), in terms of
+their equality accepting versions.
+
+    LeqAdditiveCancelationRR:
+      a + c ≤ b + c
+      ─────────────
+      a ≤ b
+    LeqAdditiveCancelationLL:
+      c + a ≤ c + b
+      ─────────────
+      a ≤ b
+    LeqAdditiveCancelationRL:
+      a + c ≤ c + b
+      ─────────────
+      a ≤ b
+    LeqAdditiveCancelationLR:
+      c + a ≤ b + c
+      ─────────────
+      a ≤ b
+    LeqMultiplicativeCancelationRR:
+      ac ≤ bc
+      ─────────────
+      a ≤ b
+    LeqMultiplicativeCancelationLL:
+      ca ≤ cb
+      ─────────────
+      a ≤ b
+    LeqMultiplicativeCancelationRL:
+      ac ≤ cb
+      ─────────────
+      a ≤ b
+    LeqMultiplicativeCancelationLR:
+      ca ≤ bc
+      ─────────────
+      a ≤ b
+
+These are known as the cancelation laws, and are essentially the inequality
+version, of those stated for equal.
+
+    LeqNegation:
+      ¬(a ≤ b)
+      ────────
+      a > b
+
+This allows us to define the _negation_ of less than or equals to.
+
+---
+
+Unfortunately, transitivity is not reductive.
+
+### Other relational definitions
+
+Non-equality is defined by:
+
+    NeqDefinition:
+      a ≠ b
+      ────────
+      ¬(a = b)
+
+## Expression reduction rules
+
+We might want to have expression reduction rules beyond the basic const
+folding. This would allow certain symbolic comparation to improve.
+
+    DistributiveMultiplicationLhs:
+      c(a + b) ↦ ca + cb
+    DistributiveMultiplicationRhs:
+      (a + b)c ↦ ca + cb
```

Comment:
Assuming that I know what I'm talking about, what matters is that both rules produce the same form. Otherwise, one will need an extra step to be able to simplify further (a step which is problematic because it is not reductive).